src/ImageSharp/Formats/Tiff/Compression/Decompressors/TiffLzwDecoder.cs

```diff
@@ -0,0 +1,255 @@
+       *
+       * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+       * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+       * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+       * DISCLAIMED.IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+       * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+       * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+       * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+       * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+       * OR TORT(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+       * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+   */
+
+    /// <summary>
+    /// Decompresses and decodes data using the dynamic LZW algorithms, see TIFF spec Section 13.
+    /// </summary>
+    internal sealed class TiffLzwDecoder
+    {
+        /// <summary>
+        /// The stream to decode.
+        /// </summary>
+        private readonly Stream stream;
+
+        /// <summary>
+        /// As soon as we use entry 4094 of the table (maxTableSize - 2), the lzw compressor write out a (12-bit) ClearCode.
+        /// At this point, the compressor reinitializes the string table and then writes out 9-bit codes again.
+        /// </summary>
+        private const int ClearCode = 256;
+
+        /// <summary>
+        /// End of Information.
+        /// </summary>
+        private const int EoiCode = 257;
+
+        /// <summary>
+        /// Minimum code length of 9 bits.
+        /// </summary>
+        private const int MinBits = 9;
+
+        /// <summary>
+        /// Maximum code length of 12 bits.
+        /// </summary>
+        private const int MaxBits = 12;
+
+        /// <summary>
+        /// Maximum table size of 4096.
+        /// </summary>
+        private const int TableSize = 1 << MaxBits;
+
+        private readonly LzwString[] table;
+
+        private int tableLength;
+        private int bitsPerCode;
+        private int oldCode = ClearCode;
+        private int maxCode;
+        private int bitMask;
+        private int maxString;
+        private bool eofReached;
+        private int nextData;
+        private int nextBits;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TiffLzwDecoder" /> class
+        /// and sets the stream, where the compressed data should be read from.
+        /// </summary>
+        /// <param name="stream">The stream to read from.</param>
+        /// <exception cref="System.ArgumentNullException"><paramref name="stream" /> is null.</exception>
+        public TiffLzwDecoder(Stream stream)
+        {
+            Guard.NotNull(stream, nameof(stream));
+
+            this.stream = stream;
+            this.table = new LzwString[TableSize];
```

Comment:
Any idea why this isn't pulled from the pool?

Reply:
Because I was unable to change the `LzwString` to a struct, i cannot use the memory allocator here.

Reply:
Cool. I'll leave a note when I do the other change for you.